### ISSUE TITLE
float conversion fails on certain substrings

### ISSUE THREAD

**Issue by mrtzh** [JuliaLang/julia#9781]
Converting `SubString` to `Float64` using `float` fails on some valid inputs:

    julia> float(split("0,1\n",",")[2])
    ERROR: ArgumentError("float64(String): invalid number format")
     in float64 at string.jl:1594
     in float at string.jl:1607

In contrast to:

    julia> float("1\n")
    1.0
    julia> split("0,1\n",",")[2]
    "1\n"

**Comment by IainNZ** [comment]
Problem seems to be here:
https://github.com/JuliaLang/julia/blob/3f4be475d112fa6b70db9300b03dfa7b8e116dac/base/string.jl#L1601-L1602
this returns false

**Comment by IainNZ** [comment]
Ah...

```
julia> x = split("0,1\n",",")
2-element Array{SubString{ASCIIString},1}:
 "0"
 "1\n"

julia> dump(x[1])
SubString{ASCIIString} "0"

julia> x[1].offset
0

julia> x[1].endof
1

julia> x[2].offset
2

julia> x[2].endof
2  # What?
```

**Comment by IainNZ** [comment]
Something seems fishy here: https://github.com/JuliaLang/julia/blob/3f4be475d112fa6b70db9300b03dfa7b8e116dac/src/builtins.c#L749-L756

**Comment by vtjnash** [comment]
cross-link https://github.com/JuliaLang/julia/pull/9487

**Comment by IainNZ** [comment]
I believe fixed by https://github.com/JuliaLang/julia/pull/9783

**Comment by mrtzh** [comment]
BTW, I still have this problem on julia 0.3.7 (Ubuntu).

**Comment by pao** [comment]
@mrtzh, the fix wasn't backported to the 0.3 branch; I've asked in that bug if it's a good candidate. Thanks!

**Comment by pao** [comment]
@mrtzh, the fix has been backported to the release-0.3 branch. If you can build Julia, you can test that now and let us know if it's fixed for you; otherwise, barring any problems it should be available in Julia 0.3.9. Thanks!
